crypto/fipsmodule/ec/p384.c

```diff
@@ -503,6 +503,479 @@ static int ec_GFp_nistp384_cmp_x_coordinate(const EC_GROUP *group,
+  fiat_p384_from_generic(p_pre_comp[0][2], &p->Z);
+
+  // Compute tmp = [2]P.
+  fiat_p384_point_double(tmp[0], tmp[1], tmp[2],
+                         p_pre_comp[0][0], p_pre_comp[0][1], p_pre_comp[0][2]);
+
+  // Generate the remaining 15 multiples of P.
+  for (size_t i = 1; i < FIAT_P384_MUL_TABLE_SIZE; i++) {
+    fiat_p384_point_add(p_pre_comp[i][0], p_pre_comp[i][1], p_pre_comp[i][2],
+                        tmp[0], tmp[1], tmp[2], 0 /* mixed */,
+                        p_pre_comp[i - 1][0], p_pre_comp[i - 1][1],
+                        p_pre_comp[i - 1][2]);
+  }
+
+  // Recode the scalar.
+  int8_t rnaf[77] = {0};
+  fiat_p384_mul_scalar_rwnaf(rnaf, scalar->bytes);
+
+  // Initialize the accumulator |res| with the table entry corresponding to
+  // the most significant digit of the recoded scalar (note that this digit
+  // can't be negative).
+  int8_t idx = rnaf[76] >> 1;
+  fiat_p384_select_point(res, idx, p_pre_comp, FIAT_P384_MUL_TABLE_SIZE);
+
+  // Process the remaining digits of the scalar.
+  for (size_t i = 75; i < 76; i--) {
+    // Double |res| 5 times in each iteration.
+    for (size_t j = 0; j < FIAT_P384_SCALAR_RADIX; j++) {
+      fiat_p384_point_double(res[0], res[1], res[2], res[0], res[1], res[2]);
+    }
+
+    int8_t d = rnaf[i];
+    // is_neg = (d < 0) ? 1 : 0
+    int8_t is_neg = (d >> 7) & 1;
+    // d = abs(d)
+    d = (d ^ -is_neg) + is_neg;
+
+    idx = d >> 1;
+
+    // Select the point to add, in constant time.
+    fiat_p384_select_point(tmp, idx, p_pre_comp, FIAT_P384_MUL_TABLE_SIZE);
+
+    // Negate y coordinate of the point tmp = (x, y); ftmp = -y.
+    fiat_p384_opp(ftmp, tmp[1]);
+    // Conditionally select y or -y depending on the sign of the digit |d|.
+    fiat_p384_cmovznz(tmp[1], is_neg, tmp[1], ftmp);
+
+    // Add the point to the accumulator |res|.
+    fiat_p384_point_add(res[0], res[1], res[2], res[0], res[1], res[2],
+                        0 /* mixed */, tmp[0], tmp[1], tmp[2]);
+
+  }
+
+  // Conditionally subtract P if the scalar is even, in constant-time.
+  // First, compute |tmp| = |res| + (-P).
+  fiat_p384_copy(tmp[0], p_pre_comp[0][0]);
+  fiat_p384_opp(tmp[1], p_pre_comp[0][1]);
+  fiat_p384_copy(tmp[2], p_pre_comp[0][2]);
+  fiat_p384_point_add(tmp[0], tmp[1], tmp[2], res[0], res[1], res[2],
+                      0 /* mixed */, tmp[0], tmp[1], tmp[2]);
+
+  // Select |res| or |tmp| based on the |scalar| parity, in constant-time.
+  fiat_p384_cmovznz(res[0], scalar->bytes[0] & 1, tmp[0], res[0]);
+  fiat_p384_cmovznz(res[1], scalar->bytes[0] & 1, tmp[1], res[1]);
+  fiat_p384_cmovznz(res[2], scalar->bytes[0] & 1, tmp[2], res[2]);
+
+  // Copy the result to the output.
+  fiat_p384_to_generic(&r->X, res[0]);
+  fiat_p384_to_generic(&r->Y, res[1]);
+  fiat_p384_to_generic(&r->Z, res[2]);
+}
+
+// Include the precomputed table for the based point scalar multiplication.
+#include "p384_table.h"
+
+// Multiplication of the base point G of P-384 curve with the given scalar.
```

Comment:
I added to the written analysis a section about the comb method in this section. We can discuss further.